pkg/cookies/csrf.go

```diff
@@ -0,0 +1,180 @@
+package cookies
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/encryption"
+	"github.com/vmihailenco/msgpack/v4"
+)
+
+var now = time.Now
+
+// CSRF manages various nonces stored in the CSRF cookie during the initial
+// authentication flows.
+type CSRF struct {
+	// State holds the OAuth2 state parameter's nonce component set in the
+	// initial authentication request and mirrored back in the callback
+	// redirect from the IdP for CSRF protection.
+	State []byte `msgpack:"s,omitempty"`
+
+	// Nonce holds the OIDC nonce parameter used in the initial authentication
+	// and then set in all subsequent OIDC ID Tokens as the nonce claim. This
+	// is used to mitigate reply attacks.
+	Nonce []byte `msgpack:"n,omitempty"`
+
+	cookieOpts *options.Cookie
+}
+
+// NewCSRF creates a CSRF with random nonces
+func NewCSRF(opts *options.Cookie) (*CSRF, error) {
+	state, err := encryption.Nonce()
+	if err != nil {
+		return nil, err
+	}
+	nonce, err := encryption.Nonce()
+	if err != nil {
+		return nil, err
+	}
+
+	return &CSRF{
+		State: state,
+		Nonce: nonce,
+
+		cookieOpts: opts,
+	}, nil
+}
+
+// HashState returns the hash of the OAuth state nonce
+func (c CSRF) HashState() string {
```

Comment:
Wondering if we should be more explicit with this and the Nonce and call them `OAuth2State` and `OIDCNonce`

Eg `csrf.HashOAuth2State`, wdyt?

Reply:
Yeah, I like it. Makes it more clear the different nonces -- I made the switch 👍